### PR TITLE
make nodecontroller delete terminating pods on 1.0 nodes

### PR DESCRIPTION
**Pull request by mikedanese**
TODO: make sure it actually works and then write tests

### CODE REVIEW

**Comment by smarterclayton** [comment]
Other than having one more cache of pods, seems accurate.

**Comment by mikedanese** [comment]
This works but exposed an issue with mirror pods. Filed #15960

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -238,6 +289,48 @@ func (nc *NodeController) getCondition(status *api.NodeStatus, conditionType api
 	return nil
 }
 
+func (nc *NodeController) checkPod(obj interface{}) {
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: this function needs a name that (1) implies it sometimes deletes pods, (2) if possible tries to describe the conditions under which it deletes pods. If (2) is not possible to express concisely you could just call the function maybeDeletePod()

also, please add a function-level comment explaining the reason for each of the deletion circumstances, or put inline comments for each of the blocks where you call nc.forcefullyDeletePod()

File: pkg/controller/node/nodecontroller.go [hunk 2]
```diff
@@ -238,6 +289,48 @@ func (nc *NodeController) getCondition(status *api.NodeStatus, conditionType api
...
+	}
+
+	if len(pod.Spec.NodeName) == 0 {
+		nc.forcefullyDeletePod(pod)
+		return
+	}
+
+	nodeObj, found, err := nc.nodeStore.GetByKey(pod.Spec.NodeName)
+	if err != nil {
+		// what do
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: random comment left here

File: pkg/controller/node/nodecontroller.go [hunk 3]
```diff
@@ -111,7 +125,8 @@ func NewNodeController(
 	nodeStartupGracePeriod time.Duration,
 	nodeMonitorPeriod time.Duration,
 	clusterCIDR *net.IPNet,
-	allocateNodeCIDRs bool) *NodeController {
+	allocateNodeCIDRs bool,
+	resyncPeriod controller.ResyncPeriodFunc) *NodeController {
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: why did you make this a parameter if you only ever call this function with the same value (controller.NoResyncPeriodFunc) ?

File: pkg/controller/node/nodecontroller.go [hunk 4]
```diff
@@ -238,6 +289,48 @@ func (nc *NodeController) getCondition(status *api.NodeStatus, conditionType api
...
+	nodeObj, found, err := nc.nodeStore.GetByKey(pod.Spec.NodeName)
+	if err != nil {
+		// what do
+		util.HandleError(err)
+		nc.podController.Requeue(pod)
+		return
+	}
+
+	if !found {
+		nc.forcefullyDeletePod(pod)
```
Comment (davidopp) [inline pkg/controller/node/nodecontroller.go]: I could have sworn we already had code somewhere that deleted pods that are on nodes that no longer exist, but it wasn't in the GC controller so maybe I'm just hallucinating?

**Comment by bgrant0607** [comment]
cc @smarterclayton

File: pkg/controller/node/nodecontroller.go (hunk 4 above)
Comment (mikedanese) [inline pkg/controller/node/nodecontroller.go]: You are hallucinating. It's one of your left to do bullets.

https://github.com/kubernetes/kubernetes/issues/7660#issuecomment-146079112

Comment (smarterclayton) [inline pkg/controller/node/nodecontroller.go]: We don't - we opened an issue to add it.

On Tue, Oct 20, 2015 at 5:19 PM, David Oppenheimer <notifications@github.com

File: pkg/controller/node/nodecontroller.go [hunk 5]
```diff
@@ -238,6 +288,59 @@ func (nc *NodeController) getCondition(status *api.NodeStatus, conditionType api
...
+		return
+	}
+
+	// TODO(mikedanese): this last if block can be remove when we no longer
+	// guarantee backwards compatibility of master API to kubelets with
+	// versions less than 1.1.0
+	node := nodeObj.(*api.Node)
+	if strings.HasPrefix(node.Status.NodeInfo.KubeletVersion, "v1.0") {
+		// delete terminating pods that have been scheudled on
```
Comment (bgrant0607) [inline pkg/controller/node/nodecontroller.go]: typo: scheduled

**Comment by mikedanese** [comment]
@k8s-bot ok to test

**Comment by davidopp** [comment]
LGTM

**Comment by ihmccreery** [comment]
@mikedanese Can you please file an issue under `kind/workaround-removal` for removing this when we drop support for `v1.0`?  Thanks!

**Comment by mikedanese** [comment]
We want this forever as well

**Comment by ihmccreery** [comment]
Ack.  Then just an issue for the last block?

**Pull request merged**